riscv/src/continuations/memory_merkle_tree.rs

```diff
@@ -30,38 +51,49 @@ fn hash_cap0<T: FieldElement>(data1: &[T; 4], data2: &[T; 4]) -> [T; 4] {
     poseidon_gl(&buffer)
 }
 
-impl<T: FieldElement, const N_LEAVES_LOG: usize, const WORDS_PER_PAGE: usize>
-    MerkleTree<T, N_LEAVES_LOG, WORDS_PER_PAGE>
+/// Takes the ordered list of node indices in one level, and updates the list
+/// in-place with the indices of the parent nodes in the level above.
+fn level_up(ordered_indices: &mut Vec<usize>) {
+    // I tried implementing this as a single dedup_by_key() call, but it doesn't
+    // invoke the closure in case of a single element, so it is not updated.
+    for i in ordered_indices.iter_mut() {
+        *i >>= 1;
+    }
+    ordered_indices.dedup();
+}
+
+impl<T: FieldElement, const N_LEVELS: usize, const WORDS_PER_PAGE: usize>
+    MerkleTree<T, N_LEVELS, WORDS_PER_PAGE>
 {
     /// Build a new Merkle tree starting from an all-zero memory.
     pub fn new() -> Self {
-        let zero_page = [T::zero(); WORDS_PER_PAGE];
-        let mut hash = Self::hash_page(&zero_page);
-        let mut hashes = vec![vec![]; N_LEAVES_LOG + 1];
+        Self::assert_field();
 
-        assert!(usize::BITS >= N_LEAVES_LOG as u32);
-        for level in (0..=N_LEAVES_LOG).rev() {
-            hashes[level] = vec![hash; 1 << level];
-            hash = hash_cap0(&hash, &hash);
+        Self {
+            hashes: HashMap::new(),
+            data: HashMap::new(),
+            default_hashes: Self::default_hashes_per_level(),
+            zero_page: [T::zero(); WORDS_PER_PAGE],
         }
-
-        assert_eq!(hashes[0].len(), 1);
-
-        let data = vec![zero_page; 1 << N_LEAVES_LOG];
-
-        Self { hashes, data }
     }
 
     /// The root hash of an empty Merkle tree.
-    /// This is equivalent to `Self::new().root_hash()`, but more memory-efficient,
-    /// because it doesn't materialize the tree.
+    ///
+    /// This is equivalent to `Self::new().root_hash()`.
```

Comment:
I think this function can go now :)

Outside tests, it is only used in `default_input()`. There, we can remove the `if accessed_pages.is_empty()` branching and always run the general algorithm, which should be fast now!

Reply:
Gone.